### ISSUE TITLE
Try to support installation of specific versions of plugins

### ISSUE THREAD

**Issue by buep** [jenkinsci/configuration-as-code-plugin#104]
For now we only support latest - let's try recommendation by @ndeloof in this comment https://github.com/jenkinsci/configuration-as-code-plugin/pull/43#issuecomment-364916028

**Comment by steven-foster** [comment]
would this support installing from http? we use some internally developed plugins that are not hosted in the update center (they wouldn't be useful to anybody)

**Comment by ndeloof** [comment]
@steven-foster the goal here is to improve plugins and dependency management, so cherry-picking binaries from outside update center without metadata is out of scope imho.

**Comment by daniel-beck** [comment]
@ndeloof @steven-foster 
The solution to that should probably involve being able to configure additional update sites, like UpdateSites Manager Plugin does. Coupled with something like Juseppe it's a clean solution using Jenkins builtin functionality.

**Comment by ndeloof** [comment]
@daniel-beck plugin management need indeed to be updated in many places.
http://updates.jenkins-ci.org/download/plugins/ let us download arbitrary plugin version but not metadata about them (some time ago @ydubreuil opened a PR to get them published as well for this exact same purpose)

**Comment by buep** [comment]
@steven-foster in case you don't get your plugin installation use-case to work, please make another issue for future work of maybe supporting your situation with self-made non-public plugins.
I know a few of our customers that have such as well, so it could be something to consider for future versions.

But it is out of scope on this issues.

**Comment by Cinderhaze** [comment]
Related to the above (hosting your own update sites), lets make sure this can still work in air-gapped environments!

**Comment by ewelinawilkosz** [comment]
@ndeloof I added 'Size' to help with planning and prediction
I follow https://www.praqma.com/stories/a-pragmatic-workflow/ where 'Size 4 - too big' simply means it takes more than a day (do you agree it does) - we can divide it to smaller <= 1 day issues if you think it makes sense

**Comment by ndeloof** [comment]
We could rely on `UpdateCenter.InstallationJob` to install plugin from a specific version, but UpdateSite unfortunately only expose a single version for each plugin (despite the download site has all of them, at least https://updates.jenkins.io/download/plugins/ does)

**Comment by ndeloof** [comment]
the sole option I have in mind is to assume update centers all use the `download/plugins/shortname/version` URL structure, so we can build plugin URL - but doing so we would miss the artifact checksum to check download was fine

**Comment by ewelinawilkosz** [comment]
@ndeloof I'm wondering what's your opinion about merging #43 - it would become part of first official release, but since you mentioned its _Only inspirational so far_ I want to know if it makes sense

**Comment by ndeloof** [comment]
@ewelinawilkosz I've been working on my own version of it here : https://github.com/ndeloof/configuration-as-code-plugin/tree/plugins
which implement the proposed JEP-201 update (https://github.com/jenkinsci/jep/pull/59)
please note this require https://github.com/jenkins-infra/update-center2/pull/192

**Comment by ndeloof** [comment]
see https://github.com/jenkinsci/configuration-as-code-plugin/pull/175

**Comment by ewelinawilkosz** [comment]
@ndeloof if I remember right current installation is not the final one, and we need to monitor https://github.com/jenkins-infra/update-center2/pull/197 
can we close this issue (after all we support the specific version install) and take necessary steps once the PR is merged in infra?

**Comment by ndeloof** [comment]
yes indeed, I think we can close this one and create a follow-up once we have an alternative to offer

**Comment by deiga** [comment]
@ndeloof Since the dependant issues seem solved, are there plans to continue this work?

**Comment by casz** [comment]
No, this feature was removed as of v1.8

**Comment by casz** [comment]
https://jenkins.io/projects/gsoc/2019/plugin-installation-manager-tool-cli/
